app/shared/app-data/src/commonMain/kotlin/domain/torrent/service/TorrentServiceConnection.kt

```diff
@@ -0,0 +1,259 @@
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.transformLatest
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.yield
+import me.him188.ani.utils.coroutines.childScope
+import me.him188.ani.utils.logging.debug
+import me.him188.ani.utils.logging.error
+import me.him188.ani.utils.logging.logger
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.EmptyCoroutineContext
+
+/**
+ * Torrent 服务与 APP 通信接口. [T] 为通信接口的类型
+ *
+ * 此接口仅负责与 Torrent 服务的通信, 启动与终止服务的逻辑可能需要在 接口的实现类(implementations) 或其他外部实现.
+ */
+interface TorrentServiceConnection<T : Any> {
+    /**
+     * 当前服务是否已连接.
+     *
+     * 若变为 `false`, 则服务通信接口将变得不可用, 接口的实现类(implementations) 可能需要重启服务, 例如 [LifecycleAwareTorrentServiceConnection].
+     */
+    val connected: StateFlow<Boolean>
+
+    /**
+     * 获取通信接口. 如果服务未连接, 则会挂起直到服务连接成功.
+     *
+     * 这个函数是线程安全的.
+     */
+    suspend fun getBinder(): T
+}
+
+/**
+ * 通过 [Lifecycle] 来约束与服务的连接状态, 保证了:
+ * - 在 [RESUMED][Lifecycle.State.RESUMED] 状态下, 根据[文档](https://developer.android.com/guide/components/activities/activity-lifecycle#onresume), APP 被视为在前台.
+ * 服务未连接或终止, 则会立刻启动或重启服务保证可用性.
+ * - 在 [CREATED][Lifecycle.State.CREATED] 和 [STARTED][Lifecycle.State.STARTED] 状态下,
+ * 若服务终止, 不会立刻重启服务, 直到再次进入 [RESUMED][Lifecycle.State.RESUMED] 状态.
+ *
+ * 实现细节:
+ *
+ * @param starter 启动服务并返回[服务通信对象][T]接口, 若返回 null 代表启动失败.
+ * @param parentCoroutineContext 执行内部逻辑的协程上下文.
+ */
+class LifecycleAwareTorrentServiceConnection<T : Any>(
+    parentCoroutineContext: CoroutineContext = EmptyCoroutineContext,
+    private val lifecycle: Lifecycle,
+    private val starter: TorrentServiceStarter<T>,
+) : TorrentServiceConnection<T> {
+    private val logger = logger(this::class.simpleName ?: "LifecycleAwareTorrentServiceConnection")
+
+    private val scope = parentCoroutineContext.childScope()
+
+    private val binderDeferred = MutableStateFlow(CompletableDeferred<T>())
+
+    // read at `onServiceDisconnected`, set by `startLifecycleLoop`
+    private val isAtForeground: MutableStateFlow<Boolean> = MutableStateFlow(false)
+    private val isServiceConnected: MutableStateFlow<Boolean> = MutableStateFlow(false)
+    private val lock = Mutex()
+    private val startServiceFlow = MutableStateFlow(Any())
+
+    private val lifecycleLoopLock = SynchronizedObject()
+    private var started = false
+
+    override val connected: StateFlow<Boolean> = isServiceConnected
+
+    fun startLifecycleLoop() {
+        if (started) return
+
+        synchronized(lifecycleLoopLock) {
+            if (started) return
+
+            scope.launch(CoroutineName("TorrentServiceConnection - RepeatOnResume")) {
+                lifecycle.repeatOnLifecycle(Lifecycle.State.RESUMED) {
+                    isAtForeground.value = true
+                    try {
+                        startServiceFlow.collectLatest {
```

Comment:
现在的这个实现仍然很复杂, 因为仍然有一些 mutable variable, 和多个线程能同时操作的 `binderDeferred`. 

建议将所有 (即一切的) `binderDeferred` 写入和决定何时重连的逻辑都放到这个 collect 内. `onServiceDisconnected` 只需要一行代码更新 flow 当前服务是否已连接. 由 collect 来综合 lifecycle + isConnected 决定是否重连.

```suggestion
                    try {
                        isServiceConnected.map { connected ->
                            // 这里没有使用 mapLatest, 因为当 connected 为 false 时我们会尝试重连并导致 connected 变为 true, mapLatest 会 cancel 掉连接协程导致可能出问题. 
                            if (connected) return
                            // service 未连接
```

不需要记录 `isAtForeground`, 因为只有这个 lifecycle loop 会处理 foreground 事情. 非 foreground 后协程会立即被 repeatOnLifecycle  cancel.